### PR TITLE
Single-stencil reproducer

### PR DESCRIPTION
**Pull request by oelbert**
## Purpose

This PR updates the external profiler to extract code and fields from a single stencil during a dycore run

Remove the sections below which do not apply.

## Code changes:

- profiler/README.MD: updated to include new command-line argument, added a todo list
- profiler/external_profiler.py: added `call_number` argument to specify which compiled version to extract.
- profiler/tools/stencil_reproducer.py: `field_serialization` now searches all directories for the cached stencil call instead of looking in a specific directory. Also added return values to exit early if the pickled stencil data already exists. `collect_stencil_candidate ` now takes a `call_number` argument to specify which version of the stencil to extract, and adds the mpi rank to the extracted stencil directory.

## Checklist
Before submitting this PR, please make sure:

- [x] You have followed the coding standards guidelines established at [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--BD7zigBMAhMZAPkeNENeuU2UAg-IlsYffZgTwyKEylty7NhY).
- [ ] Docstrings and type hints are added to new and updated routines, as appropriate
- [x] All relevant documentation has been updated or added (e.g. README, CONTRIBUTING docs)
- [ ] Unit tests are added or updated for non-stencil code changes

### CODE REVIEW

File: profiler/tools/stencil_reproducer.py [hunk 1]
```diff
@@ -33,18 +38,24 @@ def field_serialization(frame, event, args):
...
-                            f"{stencil_info[1]}/data/{prefix}_{arg_key}.npz",
+                            pickle_file,
                             arg_value.data,
                         )
                     else:
                         scalars[arg_key] = arg_value
                 scalar_file = f"{stencil_info[1]}/data/{prefix}_scalars.pickled"
                 with open(scalar_file, "wb") as handle:
                     pickle.dump(scalars, handle, protocol=pickle.HIGHEST_PROTOCOL)
+    return 0
```
Comment (FlorianDeconinck) [inline profiler/tools/stencil_reproducer.py:55]: Why a return here, it isn't used

File: profiler/external_profiler.py [hunk 2]
```diff
@@ -31,14 +31,20 @@ def parse_args():
...
+    )
     return parser.parse_known_args()
 
 
 def profile_hook(frame, event, args):
     if cmd_line_args.nvtx and nvtx_markings.mark is not None:
         nvtx_markings.mark(frame, event, args)
     if cmd_line_args.stencil and stencil_reproducer.field_serialization is not None:
-        stencil_reproducer.field_serialization(frame, event, args)
+        _ = stencil_reproducer.field_serialization(frame, event, args)
```
Comment (FlorianDeconinck) [inline profiler/external_profiler.py:47]: See below

**Review by FlorianDeconinck — APPROVED** [review]
LGTM, just a small note

**Comment by oelbert** [comment]
launch jenkins

**Comment by rheacangeo** [comment]
launch jenkins

**Comment by rheacangeo** [comment]
Jenkins is behaving really oddly, reproducibly, but not from master. And this doesn't change any code that touches the tests. So I recommend merging in master into the branch to ensure there's no strangeness coming out of the merge commit.

**Comment by rheacangeo** [comment]
Now daint is running jobs extremely slowly and you ran into the time limit.

**Comment by oelbert** [comment]
launch jenkins

**Pull request merged**